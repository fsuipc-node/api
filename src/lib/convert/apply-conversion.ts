import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
