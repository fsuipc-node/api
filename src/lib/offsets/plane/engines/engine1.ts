import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine1: OffsetList = {
  engine1ThrottleLever: new Offset({
    value: 0x88C,
    name: 'engine1ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1PropLever: new Offset({
    value: 0x88E,
    name: 'engine1PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1MixtureLever: new Offset({
    value: 0x890,
    name: 'engine1MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1StarterSwitchPosition: new Offset({
    value: 0x892,
    name: 'engine1StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine1 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1Firing: new Offset({
    value: 0x894,
    name: 'engine1Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1N2: new Offset({
    value: 0x896,
    name: 'engine1N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1N1: new Offset({
    value: 0x898,
    name: 'engine1N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1PropRPM: new Offset({
    value: 0x898,
    name: 'engine1PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1ThrottleLeverControl: new Offset({
    value: 0x89A,
    name: 'engine1ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine1 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1FuelFlowLbHourSSL: new Offset({
    value: 0x8A0,
    name: 'engine1FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1AntiIce: new Offset({
    value: 0x8B2,
    name: 'engine1AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine1 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilTemp: new Offset({
    value: 0x8B8,
    name: 'engine1OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilPres: new Offset({
    value: 0x8BA,
    name: 'engine1OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1PressureRatio: new Offset({
    value: 0x8BC,
    name: 'engine1PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine1 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1EGT: new Offset({
    value: 0x8BE,
    name: 'engine1EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1MP: new Offset({
    value: 0x8C0,
    name: 'engine1MP',
    category: OffsetCategory.ENGINE,
    description: 'engine1 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1RPMScaler: new Offset({
    value: 0x8C8,
    name: 'engine1RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine1 RPM Scaler - props: used to calculate RPM',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilQuantity: new Offset({
    value: 0x8D0,
    name: 'engine1OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1Vibration: new Offset({
    value: 0x8D4,
    name: 'engine1Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine1 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1HydPres: new Offset({
    value: 0x8D8,
    name: 'engine1HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1HydQuantity: new Offset({
    value: 0x8DC,
    name: 'engine1HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic quantity - PSI',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1CHT: new Offset({
    value: 0x8E8,
    name: 'engine1CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine1ITT: new Offset({
    value: 0x8F0,
    name: 'engine1ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1Torque: new Offset({
    value: 0x8F4,
    name: 'engine1Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine1 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1FuelPres: new Offset({
    value: 0x8F8,
    name: 'engine1FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1FuelUsedSinceStart: new Offset({
    value: 0x90C,
    name: 'engine1FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine1 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelElapsedTime: new Offset({
    value: 0x910,
    name: 'engine1FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelFlowLbHour: new Offset({
    value: 0x918,
    name: 'engine1FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TorqueFtLb: new Offset({
    value: 0x920,
    name: 'engine1TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
};
