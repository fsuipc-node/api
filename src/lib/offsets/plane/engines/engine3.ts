import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine3: OffsetList = {
  engine3ThrottleLever: new Offset({
    value: 0x9BC,
    name: 'engine3ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine3PropLever: new Offset({
    value: 0x9BE,
    name: 'engine3PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3MixtureLever: new Offset({
    value: 0x9C0,
    name: 'engine3MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3StarterSwitchPosition: new Offset({
    value: 0x9C2,
    name: 'engine3StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine3 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine3Firing: new Offset({
    value: 0x9C4,
    name: 'engine3Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine3N2: new Offset({
    value: 0x9C6,
    name: 'engine3N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3N1: new Offset({
    value: 0x9C8,
    name: 'engine3N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 3 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3PropRPM: new Offset({
    value: 0x9C8,
    name: 'engine3PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3ThrottleLeverControl: new Offset({
    value: 0x9CA,
    name: 'engine3ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine3 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine3FuelFlowLbHourSSL: new Offset({
    value: 0x9D0,
    name: 'engine3FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine3 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine3AntiIce: new Offset({
    value: 0x9E2,
    name: 'engine3AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine3 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine3OilTemp: new Offset({
    value: 0x9E8,
    name: 'engine3OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine3 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine3OilPres: new Offset({
    value: 0x9EA,
    name: 'engine3OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine3 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine3PressureRatio: new Offset({
    value: 0x9EC,
    name: 'engine3PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine3 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.Int16,
    permission: 'r',
  }),
  engine3EGT: new Offset({
    value: 0x9EE,
    name: 'engine3EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine3 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine3MP: new Offset({
    value: 0x9F0,
    name: 'engine3MP',
    category: OffsetCategory.ENGINE,
    description: 'engine3 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine3RPMScaler: new Offset({
    value: 0x9F8,
    name: 'engine3RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine3 RPM Scaler - props: used to calculate RPM',
    type: Type.Int16,
    permission: 'r',
  }),
  engine3OilQuantity: new Offset({
    value: 0xA00,
    name: 'engine3OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine3 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3Vibration: new Offset({
    value: 0xA04,
    name: 'engine3Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine3 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3HydPres: new Offset({
    value: 0xA08,
    name: 'engine3HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine3 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3HydQuantity: new Offset({
    value: 0xA0C,
    name: 'engine3HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine3 hydraulic quantity - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3CHT: new Offset({
    value: 0xA18,
    name: 'engine3CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine3 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine3ITT: new Offset({
    value: 0xA20,
    name: 'engine3ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine3 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3Torque: new Offset({
    value: 0xA24,
    name: 'engine3Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine3 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3FuelPres: new Offset({
    value: 0xA28,
    name: 'engine3FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine3 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.Int32,
    permission: 'r',
  }),
  engine3FuelUsedSinceStart: new Offset({
    value: 0xA3C,
    name: 'engine3FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine3 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine3FuelElapsedTime: new Offset({
    value: 0xA40,
    name: 'engine3FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine3 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine3FuelFlowLbHour: new Offset({
    value: 0xA48,
    name: 'engine3FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine3 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine3TorqueFtLb: new Offset({
    value: 0xA50,
    name: 'engine3TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine3 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
};
