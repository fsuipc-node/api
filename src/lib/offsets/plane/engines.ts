import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

import { engine1 } from './engines/engine1';
import { engine2 } from './engines/engine2';
import { engine3 } from './engines/engine3';
import { engine4 } from './engines/engine4';

export const engines: OffsetList = {
  engineType: new Offset({
    value: 0x609,
    name: 'engineType',
    category: OffsetCategory.ENGINE,
    description: 'engine type',
    convert: 'engineType',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  hasMixtureControl: new Offset({
    value: 0x780,
    name: 'hasMixtureControl',
    category: OffsetCategory.ENGINE,
    description: 'has mixture control',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.ENGINE,
    description: 'has carb heat',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  activeEngine: new Offset({
    value: 0x888,
    name: 'activeEngine',
    category: OffsetCategory.ENGINE,
    description: 'active engine pattern',
    type: Type.BitArray,
    length: 1,
    permission: 'rw',
  }),
  ...engine1,
  ...engine2,
  ...engine3,
  ...engine4,
};
