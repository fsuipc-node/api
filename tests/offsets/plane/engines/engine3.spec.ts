import { engine3 as offsets } from '@offsets/plane/engines/engine3';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/engines/engine3', () => {
  const offsetsTestCases = [
    { name: 'engine3ThrottleLever', value: -2048, expectedResult: -50 },
    { name: 'engine3ThrottleLever', value: 0, expectedResult: 0 },
    { name: 'engine3ThrottleLever', value: 8192, expectedResult: 50 },
    { name: 'engine3PropLever', value: -2048, expectedResult: -50 },
    { name: 'engine3PropLever', value: 0, expectedResult: 0 },
    { name: 'engine3PropLever', value: 8192, expectedResult: 50 },
    { name: 'engine3MixtureLever', value: 8192, expectedResult: 50 },
    { name: 'engine3Firing', value: 0, expectedResult: false },
    { name: 'engine3Firing', value: 1, expectedResult: true },
    { name: 'engine3N2', value: 8192, expectedResult: 50 },
    { name: 'engine3N1', value: 8192, expectedResult: 50 },
    { name: 'engine3PropRPM', value: 107374182400, expectedResult: 100 },
    { name: 'engine3FuelFlowLbHourSSL', value: 256, expectedResult: 2 },
    { name: 'engine3AntiIce', value: 0, expectedResult: false },
    { name: 'engine3AntiIce', value: 1, expectedResult: true },
    { name: 'engine3OilTemp', value: 10532, expectedResult: 90 },
    { name: 'engine3OilPres', value: 22342, expectedResult: 75 },
    { name: 'engine3PressureRatio', value: 20480, expectedResult: 2 },
    { name: 'engine3EGT', value: 14281, expectedResult: 750 },
    { name: 'engine3MP', value: 20480, expectedResult: 20 },
    { name: 'engine3OilQuantity', value: 8167, expectedResult: 50 },
    { name: 'engine3Vibration', value: 6553, expectedResult: 2 },
    { name: 'engine3HydPres', value: 64, expectedResult: 16 },
    { name: 'engine3HydQuantity', value: 8192, expectedResult: 50 },
    { name: 'engine3ITT', value: 3276800, expectedResult: 200 },
    { name: 'engine3Torque', value: 8192, expectedResult: 50 },
    { name: 'engine3FuelPres', value: 288, expectedResult: 2 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
