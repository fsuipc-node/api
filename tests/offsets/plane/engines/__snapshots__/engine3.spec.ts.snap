// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/engines/engine3 offsets list should have required properties 1`] = `
Object {
  "engine3AntiIce": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine3 anti-ice/carb heat active",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3AntiIce",
    "permission": "r",
    "type": 5,
    "value": 2530,
  },
  "engine3CHT": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 cylinder head temperature - F",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3CHT",
    "permission": "r",
    "type": 8,
    "value": 2584,
  },
  "engine3EGT": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 860 / 16384)",
    "description": "engine3 exhaust gas temperature - Trust only on jet engine - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3EGT",
    "permission": "r",
    "type": 5,
    "value": 2542,
  },
  "engine3Firing": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 3 combustion",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3Firing",
    "permission": "r",
    "type": 5,
    "value": 2500,
  },
  "engine3FuelElapsedTime": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 fuel elapsed time - h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3FuelElapsedTime",
    "permission": "r",
    "type": 9,
    "value": 2624,
  },
  "engine3FuelFlowLbHour": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 fuel elapsed time - lb/h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3FuelFlowLbHour",
    "permission": "r",
    "type": 8,
    "value": 2632,
  },
  "engine3FuelFlowLbHourSSL": Offset {
    "category": "engine",
    "convert": "{VAL} / 128",
    "description": "engine3 fuel flow (lb per hour, sea level)",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3FuelFlowLbHourSSL",
    "permission": "r",
    "type": 5,
    "value": 2512,
  },
  "engine3FuelPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 144",
    "description": "engine3 fuel pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3FuelPres",
    "permission": "r",
    "type": 6,
    "value": 2600,
  },
  "engine3FuelUsedSinceStart": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 used fuel since start - lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3FuelUsedSinceStart",
    "permission": "r",
    "type": 9,
    "value": 2620,
  },
  "engine3HydPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 4",
    "description": "engine3 hydraulic pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3HydPres",
    "permission": "r",
    "type": 6,
    "value": 2568,
  },
  "engine3HydQuantity": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine3 hydraulic quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3HydQuantity",
    "permission": "r",
    "type": 6,
    "value": 2572,
  },
  "engine3ITT": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384",
    "description": "engine3 turbine temperature - C",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3ITT",
    "permission": "r",
    "type": 6,
    "value": 2592,
  },
  "engine3MP": Offset {
    "category": "engine",
    "convert": "{VAL} / 1024",
    "description": "engine3 manifold pressure - inHg",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3MP",
    "permission": "r",
    "type": 5,
    "value": 2544,
  },
  "engine3MixtureLever": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 3 prop lever - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3MixtureLever",
    "permission": "rw",
    "type": 5,
    "value": 2496,
  },
  "engine3N1": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 3 N1 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3N1",
    "permission": "rw",
    "type": 2,
    "value": 2504,
  },
  "engine3N2": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 3 N2 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3N2",
    "permission": "rw",
    "type": 5,
    "value": 2502,
  },
  "engine3OilPres": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 55 / 16384)",
    "description": "engine3 oil pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3OilPres",
    "permission": "r",
    "type": 5,
    "value": 2538,
  },
  "engine3OilQuantity": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} / 16384 * 100)",
    "description": "engine3 oil quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3OilQuantity",
    "permission": "r",
    "type": 6,
    "value": 2560,
  },
  "engine3OilTemp": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 140 / 16384)",
    "description": "engine3 oil temp - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3OilTemp",
    "permission": "r",
    "type": 5,
    "value": 2536,
  },
  "engine3PressureRatio": Offset {
    "category": "engine",
    "convert": "{VAL} * 1.6 / 16384",
    "description": "engine3 pressure ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3PressureRatio",
    "permission": "r",
    "type": 5,
    "value": 2540,
  },
  "engine3PropLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 3 prop lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3PropLever",
    "permission": "rw",
    "type": 2,
    "value": 2494,
  },
  "engine3PropRPM": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 / 65536",
    "description": "derive RPM by multiplying by  - negative = counter-rotating propeller",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3PropRPM",
    "permission": "rw",
    "type": 2,
    "value": 2504,
  },
  "engine3RPMScaler": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 RPM Scaler - props: used to calculate RPM",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3RPMScaler",
    "permission": "r",
    "type": 5,
    "value": 2552,
  },
  "engine3StarterSwitchPosition": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3StarterSwitchPosition",
    "permission": "rw",
    "type": 5,
    "value": 2498,
  },
  "engine3ThrottleLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 3 throttle lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3ThrottleLever",
    "permission": "r",
    "type": 2,
    "value": 2492,
  },
  "engine3ThrottleLeverControl": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 throttle lever control - -4096 to +16384",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3ThrottleLeverControl",
    "permission": "rw",
    "type": 2,
    "value": 2506,
  },
  "engine3Torque": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine3 torque - trust only for helo - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3Torque",
    "permission": "r",
    "type": 6,
    "value": 2596,
  },
  "engine3TorqueFtLb": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine3 fuel elapsed time - ft lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3TorqueFtLb",
    "permission": "r",
    "type": 9,
    "value": 2640,
  },
  "engine3Vibration": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 5 / 16384)",
    "description": "engine3 vibration",
    "length": undefined,
    "mapping": undefined,
    "name": "engine3Vibration",
    "permission": "r",
    "type": 6,
    "value": 2564,
  },
}
`;
